backend/test/richard_burton/publication_index_test.exs

```diff
@@ -0,0 +1,182 @@
+defmodule RichardBurton.Publication.IndexTest do
+  @moduledoc """
+  Tests for the Publication.Index module
+  """
+
+  use RichardBurton.DataCase
+
+  alias RichardBurton.Publication
+
+  setup(_context) do
+    {:ok, publications} = Publication.Codec.from_csv("test/fixtures/data_index.csv")
+
+    ps =
+      publications
+      |> Enum.map(&insert!/1)
+      |> Publication.preload()
+      |> Publication.Codec.flatten()
+      |> Enum.map(&Map.to_list/1)
+      |> Enum.map(&Enum.map(&1, fn {k, v} -> {String.to_existing_atom(k), v} end))
+      |> Enum.map(&Map.new/1)
+
+    [publications: ps]
+  end
+
+  defp insert!(attrs) do
+    %Publication{} |> Publication.changeset(attrs) |> Repo.insert!()
+  end
+
+  defp sort(publications) do
+    Enum.sort(publications)
+  end
+
+  defp filter(publications, key, term) do
+    Enum.filter(publications, &String.contains?(to_string(&1[key]), term))
+  end
+
+  describe "all/0" do
+    test "returns all publications", context do
+      %{publications: publications} = context
+      {:ok, result} = Publication.Index.all()
+
+      assert sort(publications) == sort(result)
+    end
+  end
+
+  describe "search/1 with a single-word term present in the dataset" do
+    test "retrieves publications by original author", context do
+      %{publications: publications} = context
+
+      term = "Verissimo"
+      keyword = String.downcase(term)
+
+      assert {:ok, result, [^keyword]} = Publication.Index.search(term)
+
+      assert length(result) > 0
+      assert sort(filter(publications, :original_authors, term)) == sort(result)
```

Comment:
This is not Elixir-related but I'm throwing it anyways :)

I feel it is usually better to use literals (hardcoded) as expected values in unit testing assertions. When we build them using custom logic we're hiding what the expected result of the functions should be. In this case, I would love to see the result of doing `sort(filter(publications, :original_authors, term))` 😄 

Related readings (advocating to this way of writing the expected values)

https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices#avoid-logic-in-tests
https://canro91.github.io/2021/10/11/DontRepeatLogicInAssertions/

Reply:
I noticed that the complete expected value would include a list of 5 elements, which is too verbose. So let me show what I would do to apply what I explained in the previous comment:

```ex
  assert {:ok, result, [^keyword]} = Publication.Index.search(term)
      
  # The expected length is exactly 5 results (according to what I see in data_index.csv)
  assert length(result) == 5
      
  # Checking the whole list is too verbose, so you can check the first one using pattern matching
  # The map should contain the real data you have in data_index.csv, but I'm omitting this to keep the example short
  assert [%{original_authors: ^term, ...}| _] = result
```

An additional assertion you may want to have (or not) is the following, to confirm that every item in the result has the expected author:

```ex
  # Note that the pattern matching uses a part of the map, you can not include anything expect the `original_authors` key
  Enum.each(result, fn r -> assert %{original_authors: ^term} = r end)
```

Thinking again about this case I see you are not *duplicating the production code in tests* but you're still using some logic there, so this is just an alternative way to approach the tests, with pros and cons (to discuss in person 😸)

Reply:
Agreed, I tried to abide by that in most of the tests I've made.

These tests are fairly complex conceptually, though. Since the search is fuzzy, I have to be really careful with the arguments and expected result, so I can't assert just the length and the first one. 

The second assertion would be the way to go, It's smart to just check the relevant field. However, `term` or `keyword` cannot be used. I.e.: with the complete dataset, `ver` as search term should result in keywords [`verbo`, `verisimo`], which should in turn result in publications with "Erico Verissimo" or "Luis Fernando Verissimo" as original authors, or "Amar verbo intransitivo" as original title.

I might just hardcode the expected values for `original_authors` (or the relevant key) and assert on that. Thoughts on that?

Reply:
Should be changed in following commits, I think it's way better now 🎉 🙌